### ISSUE TITLE
Parameter 1 to MongoCollection::insert() expected to be a reference, value given

### ISSUE THREAD

**Issue by glowdan** [alcaeus/mongo-php-adapter#107]
Hi, 
When we upgrade our website from php5.5 to php7.0.6. We got some warning like this:

> PHP Warning:  Parameter 1 to MongoCollection::insert() expected to be a reference, value given in /website/helper/mongo/collection.php on line 66.

And I saw the same question occurs on[ wordpress](https://www.itnota.com/fixing-php7-compatibility-issue-w3-total-cache/). May I remove the "&" in "alcaeus/mongo-php-adapter/lib/Mongo/MongoCollection.php::insert on line 277"?

**Comment by alcaeus** [comment]
Unfortunately, `insert` and `batchInsert` and `save` can't be implemented the way they were in the legacy driver:
> The insert, batchInsert, and save methods take the first argument by reference. While the original API does not explicitely specify by-reference arguments it does add an ID field to the objects and documents given.

I had the choice:
1. have the first argument by-value: this would have allowed passing a value to the method but the array will be lacking the `_id` value that is added by the driver
2. have the first argument by-reference: this would fix the `_id` problem but the method signatures would be incompatible.

The original intent of the driver was to have the argument by-reference so that an auto-generated ID could be added. Thus, I settled for option #2, knowing that people can either handle that specific warning in an error handler or ignore it. On the other hand, having code break because you were expecting an `_id` field in your array seems a lot more severe.

TL;DR: I wouldn't recommend it.

**Comment by glowdan** [comment]
OK，I solved it. If we need insert or save data to db, we add the reference to the first param.
```php
            if (PHP_VERSION_ID > 70000 && in_array($method, ['insert', 'batchInsert', 'save'])) {
                $saveData = array_shift($param);
                $saveParams = array_shift($param);
                if (NULL==$saveParams) {
                    $saveParams = [];
                }
                $re = call_user_func_array([$collection, $method], [&$saveData, $saveParams]);
            } else {
                $re = call_user_func_array([$collection, $method], $param);
            }
```

**Comment by Ardakilic** [comment]
@glowdan can you please elaborate more ? Where did you add this patch ?

I'm using https://github.com/jenssegers/laravel-mongodb with Laravel 4.2.x along with PHP7.0, and this is the only step I'm struggling with.

**Comment by rhclayto** [comment]
@Ardakilic Check this out: https://github.com/LearningLocker/learninglocker/issues/893

**Comment by Ardakilic** [comment]
Thanks @rhclayto  , will look right away!

**Comment by Ardakilic** [comment]
Hey there, sorry for the very late reply, I can confirm @glowdan fix works flawlessly on PHP7! 🎉  I've created a PR to the base repository.
